oauth2_http/java/com/google/auth/oauth2/DefaultCredentialsProvider.java

```diff
@@ -141,6 +158,76 @@ private final File getWellKnownCredentialsFile() {
     return credentialFilePath;
   }
 
+  private boolean runningOnAppEngine() {
+    Class<?> systemPropertyClass = null;
+    try {
+      systemPropertyClass = forName(APP_ENGINE_SIGNAL_CLASS);
+    } catch (ClassNotFoundException expected) {
+      // SystemProperty will always be present on App Engine.
+      return false;
+    }
+    Exception cause = null;
+    Field environmentField;
+    try {
+      environmentField = systemPropertyClass.getField("environment");
+      Object environmentValue = environmentField.get(null);
+      Class<?> environmentType = environmentField.getType();
+      Method valueMethod = environmentType.getMethod("value");
+      Object environmentValueValue = valueMethod.invoke(environmentValue);
+      return (environmentValueValue != null);
+    } catch (NoSuchFieldException exception) {
+      cause = exception;
+    } catch (SecurityException exception) {
+      cause = exception;
+    } catch (IllegalArgumentException exception) {
+      cause = exception;
+    } catch (IllegalAccessException exception) {
+      cause = exception;
+    } catch (NoSuchMethodException exception) {
+      cause = exception;
+    } catch (InvocationTargetException exception) {
+      cause = exception;
+    }
+    throw OAuth2Utils.exceptionWithCause(new RuntimeException(String.format(
+        "Unexpcted error trying to determine if runnning on Google App Engine: %s",
+        cause.getMessage())), cause);
+  }
+
+  private GoogleCredentials tryGetAppEngineCredential() throws IOException {
+    // Checking for App Engine requires a class load, so check only once
+    if (checkedAppEngine) {
+      return null;
+    }
+    boolean onAppEngine = runningOnAppEngine();
+    checkedAppEngine = true;
+    if (!onAppEngine) {
+      return null;
+    }
+    Exception innerException = null;
+    try {
+      Class<?> credentialClass = forName(APP_ENGINE_CREDENTIAL_CLASS);
+      Constructor<?> constructor = credentialClass
+          .getConstructor(Collection.class);
+      Collection<String> emptyScopes = Collections.emptyList();
+      return (GoogleCredentials) constructor.newInstance(emptyScopes);
+    } catch (ClassNotFoundException e) {
```

Comment:
Ditto above, but including `ClassNotFoundException` and `InstantiationException`.